### PR TITLE
deploy vsix packages on build

### PR DESCRIPTION
**Pull request by jmarolf**
Currently if someone runs `Build.cmd` and then launches visual studio using `devenv /rootSuffix RoslynDev` none of the vsix packages will have been deployed.  Deployment *without opening Visual Studio* is pretty roundabout without this change.

### CODE REVIEW

**Comment by jaredpar** [comment]
Not sure I understand the problem here. The default is that Build.cmd, and more simply `msbuild Roslyn.sln`, will deploy our VSIX. Is that not happening here? The PR as authored causes us to deploy twice during build.

**Comment by jmarolf** [comment]
@jaredpar `Build.cmd` does not deploy vsixes today.  Let me fire up a fresh VM but several people internally and externally have said this is not working from them

**Comment by jaredpar** [comment]
The intent is def for it to deploy on build. I actually just fixed some bugs around that a few days ago. If it’s not then we prob need to see why DeployExtension isn’t being set in the binlog

**Comment by sharwell** [comment]
@jmarolf Have you verified that this change does not noticeably increase the time it takes to run **Build.cmd**?

File: build/scripts/build.ps1 [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #
 # This script controls the Roslyn build process. This encompasess everything from build, testing to
-# publishing of NuGet packages. The intent is to structure it to allow for a simple flow of logic 
+# publishing of NuGet packages. The intent is to structure it to allow for a simple flow of logic
```
Comment (sharwell) [inline build/scripts/build.ps1]: Lots of unrelated changes in the file will make a mess of git blame and now I have to go find the relevant change(s). ☹️

Comment (jmarolf) [reply build/scripts/build.ps1]: I can add back all the whitespace

**Comment by jmarolf** [comment]
@jaredpar Looks like we've been explicitly passing `DeployExtension = false` to every invocation of MSBuild that builds `Roslyn.sln`
https://github.com/dotnet/roslyn/blob/80ae901c654f39666130dc94458ba673154a8859/build/scripts/build.ps1#L249

Am I missing something, how is `build.cmd` deploying extensions supposed to work?

**Comment by jaredpar** [comment]
> Am I missing something, how is build.cmd deploying extensions supposed to work?

Hmmm. I can't remember why I added that. My initial reaction is that I did that as a temp local work around and accidentally checked it in and have been mindlessly maintaining it since then.

This code goes against discussions I've had with IDE team members in the past. The intent has always been that build should deploy the VSIX because that's what developers expect it to do. In the case of CI and such we should suppress it where it makes sense but overall build = I can debug the VSIX. 

Lets sync up with rest of IDE and reset what the expectation is here.

**Comment by jmarolf** [comment]
@dotnet/roslyn-ide it appears that commandline builds have not been deploying VSIXes for a while.  The only way it worked was building from Visual Studio.  I've changed it to deploy extensions by default but there is a switch so you can disable it if you want.

This change is motivated by the feedback I had gotten from contributors that were confused by this behavior.

**Comment by jmarolf** [comment]
Note: until https://github.com/dotnet/roslyn/issues/26403 is fixed the build will fail because we generate invalid VSIXes

**Comment by jmarolf** [comment]
@sharwell 

> This change adds another 37 seconds to deal with deployment.

If you pull this PR down and build it it will fail due  to #26403 so I don't know how you are getting those measurements. With PR https://github.com/dotnet/roslyn/pull/26404 no additional time should be added to deployment, (VSIX deployment will happen during solution build).  

We can have a separate discussion about build times for roslyn (can you file and issue?). If we are worried about build times we should not have build.cmd run `Build-ExtraSignArtifacts` as that is a large set of things that don't matter any developer running tests on their machine (frankly I don't know how @jaredpar was allowed to sneak that change in without everyone screaming).

https://github.com/dotnet/roslyn/blob/7c8e2e57210eecee88f7cd708880f1cd53bc70c3/build/scripts/build.ps1#L248-L251

https://github.com/dotnet/roslyn/blob/7c8e2e57210eecee88f7cd708880f1cd53bc70c3/build/scripts/build.ps1#L270-L301

**Comment by jaredpar** [comment]
> frankly I don't know how @jaredpar was allowed to sneak that change in without everyone screaming

Before there was `-build` and `-buildAll` where the latter did the Build-ExtraSignArtifacts portion. Sure `-build` was faster but it also was confusing because commands like `-pack` and `-sign` didn't work unless you had run `-buildAll` first. It was a clear functional gap and I got **plenty** of nasty-grams about that. Hence I moved to just a simple `-build` command. 

Other reason for that was the following. If you really just want to build Roslyn.sln and are upset about the overhead that `Build.cmd` provides then just run `msbuild Roslyn.sln` or use Visual Studio.

**Comment by sharwell** [comment]
> If you pull this PR down and build it it will fail due to #26403 so I don't know how you are getting those measurements. With PR #26404 no additional time should be added to deployment, (VSIX deployment will happen during solution build).

With #26404 applied, the time to build Roslyn.sln increases from 19 seconds to 54 seconds (net 35 second build time increase).

**Comment by jmarolf** [comment]
@jaredpar I think you mean `msbuild /m /v:m Roslyn.sln` but point taken

**Comment by jmarolf** [comment]
@sharwell as a compromise could we check in the change ro `build.ps1` but not the change to `build.cmd`?  That way `build.cmd` time won't regress but then at least everytime someone comes to me and says build doesn't work I can tell them to run `build.cmd -deployExtensions`.

**Comment by sharwell** [comment]
@jmarolf That would be fine with me 😄

**Comment by jasonmalinowski** [comment]
Would that create new confusion though? Would typing 'build' (note: no suffix) run one in powershell and the other in cmd?

**Comment by jmarolf** [comment]
> Would that create new confusion though? Would typing 'build' (note: no suffix) run one in powershell and the other in cmd?

`build.cmd` is located in the root and `build.ps1` is located in `build\scripts` for this reason.  Today having a folder named `build` does make autocomplete harder for powershell users, but that is orthogonal.

**Comment by jmarolf** [comment]
@sharwell any objections to this new approach?

File: build/scripts/build.ps1 [hunk 2]
```diff
@@ -246,7 +248,7 @@ function Build-Artifacts() {
         Run-MSBuild "Compilers.sln" -useDotnetBuild
     }
     elseif ($build) {
-        Run-MSBuild "Roslyn.sln" "/p:DeployExtension=false"
+        Run-MSBuild "Roslyn.sln" "/p:DeployExtension=$deployExtensions"
```
Comment (sharwell) [inline build/scripts/build.ps1]: :bulb: it may be better to omit the property than set it to true. Setting it to true removes the ability of a project to unconditionally disable deployment via properties in the project file.
